src/Catalog/Filter/LocalizedCatalogLocaleFilter.php

```diff
@@ -0,0 +1,127 @@
+<?php
+/**
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade Gally to newer versions in the future.
+ *
+ * @package   Gally
+ * @author    Gally Team <elasticsuite@smile.fr>
+ * @copyright 2022-present Smile
+ * @license   Open Software License v. 3.0 (OSL-3.0)
+ */
+
+declare(strict_types=1);
+
+namespace Gally\Catalog\Filter;
+
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Filter\SearchFilter;
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Util\QueryNameGeneratorInterface;
+use Doctrine\ORM\Query\Expr\Join;
+use Doctrine\ORM\QueryBuilder;
+use Gally\Catalog\Model\LocalizedCatalog;
+use Gally\Exception\LogicException;
+use Symfony\Component\PropertyInfo\Type;
+
+/**
+ * Filters the collection on localizedCatalogs and the locale related to localizedCatalogs.
+ * The property class that store the locale must be passed on filter properties via the key locale_property.
+ * For example:
+ * #[ApiFilter(LocalizedCatalogLocaleFilter::class, properties: ['localizedCatalogs.id' => ['locale_property' => 'locales.locale']])]
+ * Here locales.locale is the property where locales are stored.
+ */
+class LocalizedCatalogLocaleFilter extends SearchFilter
+{
+    protected function filterProperty(
+        string $property,
+        $value,
+        QueryBuilder $queryBuilder,
+        QueryNameGeneratorInterface $queryNameGenerator,
+        string $resourceClass,
+        string $operationName = null
+    ) {
+        if (
+            !$this->isPropertyEnabled($property, $resourceClass) ||
+            !$this->isPropertyMapped($property, $resourceClass)
+        ) {
+            return;
+        }
+
+        $alias = $queryBuilder->getRootAliases()[0];
+
+        $values = $this->normalizeValues((array) $value, $property);
+        $values = array_map([$this, 'getIdFromValue'], $values);
+        if (empty($values)) {
+            return;
+        }
+
+        $propertiesToFilter = $this->properties[$property] ?? [];
+        $localeProperty = $propertiesToFilter['locale_property'] ?? null;
+
+        if (null === $localeProperty) {
+            throw new LogicException("The 'locale_property' is not set in the filter configuration.");
+        }
+
+        [$localizedCatalogAlias] = $this->addJoinsForNestedProperty(
+            $property,
+            $alias,
+            $queryBuilder,
+            $queryNameGenerator,
+            $resourceClass,
+            Join::LEFT_JOIN
+        );
+
+        [$localeAlias, $localeField] = $this->addJoinsForNestedProperty(
+            $localeProperty,
+            $alias,
+            $queryBuilder,
+            $queryNameGenerator,
+            $resourceClass,
+            Join::LEFT_JOIN
+        );
```

Comment:
OK, seen the logic in \ApiPlatform\Core\Bridge\Doctrine\Orm\PropertyHelperTrait::addJoinsForNestedProperty.

You grab the returned field in the second call because you need it a few lines after that.